lib/google/ads/google_ads/deprecation.rb

```diff
@@ -0,0 +1,96 @@
+module Google
+  module Ads
+    module GoogleAds
+      class Deprecation
+        def initialize(
+          treat_deprecation_warnings_as_errors,
+          warn_on_all_deprecations
+        )
+          @treat_deprecation_warnings_as_errors = treat_deprecation_warnings_as_errors
+          @warn_on_all_deprecations = warn_on_all_deprecations
+          @silenced_sites = {}
+        end
+
+        def deprecate(deprecation)
+          return unless should_warn?
+          if @treat_deprecation_warnings_as_errors
+            raise deprecation
+          else
+            Warning.warn("#{deprecation}. Called from: #{CallerFilter.first_non_google_ads_line}")
+          end
+        end
+
+        private
+
+        # Determines if we should issue a deprecation warning, silencing
+        # on each user call site after the first warning.
+        def should_warn?
+          true if @warn_on_all_deprecations
```

Comment:
I don't think this returns since it's not the last statement in the method and it doesn't explicitly say return.